### PR TITLE
Backend unit tests

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Fixes # (issue)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/overrides/fairscale.py
Comment (Borda) [inline pytorch_lightning/overrides/fairscale.py:18]:
```suggestion
FAIRSCALE_SHARDED_AVAILABLE = _module_available('fairscale.nn.data_parallel.sharded_ddp'):
if FAIRSCALE_SHARDED_AVAILABLE:
```
maybe move the `FAIRSCALE_SHARDED_AVAILABLE` to utils init as others

File: pytorch_lightning/overrides/fairscale.py
Comment (Borda) [inline pytorch_lightning/overrides/fairscale.py:37]:
```suggestion
else:
    LightningShardedDataParallel = None
```

File: pytorch_lightning/plugins/sharded_native_amp_plugin.py
Comment (Borda) [inline pytorch_lightning/plugins/sharded_native_amp_plugin.py:22]: let's move these to `utils.__init__` and just use it here

File: pytorch_lightning/plugins/sharded_plugin.py
Comment (Borda) [inline pytorch_lightning/plugins/sharded_plugin.py:54]: this is tricky, if you assume that the first is always batch, lets define it in the method and the later are args

File: tests/plugins/test_sharded_plugin.py
Comment (Borda) [inline tests/plugins/test_sharded_plugin.py:14]:
```suggestion
from pytorch_lightning.plugins.sharded_plugin import DDPShardedPlugin
from pytorch_lightning.utilities import NATIVE_AMP_AVALAIBLE, FAIRSCALE_AVAILABLE
```
lets move it here

File: tests/plugins/test_sharded_plugin.py
Comment (Borda) [inline tests/plugins/test_sharded_plugin.py:350]:
```suggestion
@pytest.mark.skipif(not FAIRSCALE_AVAILABLE, reason="Fairscale is not available")
```

File: tests/plugins/test_sharded_plugin.py
Comment (Borda) [inline tests/plugins/test_sharded_plugin.py:363]: could the name be a bit more descriptive? :]

File: tests/plugins/test_sharded_plugin.py
Comment (Borda) [inline tests/plugins/test_sharded_plugin.py:423]: I guess this is mo related to the benchmark, right?

File: tests/plugins/test_sharded_plugin.py
Comment (Borda) [inline tests/plugins/test_sharded_plugin.py:462]:
```suggestion
    Ensures that the trained model is identical to the standard DDP implementation.
    Also checks for speed/memory regressions, we should expect always less memory but performance to fluctuate.

```

File: tests/plugins/test_sharded_plugin.py
Comment (Borda) [inline tests/plugins/test_sharded_plugin.py:454]: let's move this to the benchmark folder to other parity tests

**Review by Borda — COMMENTED** [review]
can we add docstring?

File: pytorch_lightning/trainer/connectors/precision_connector.py
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/precision_connector.py:72]: Does it work with Apex ?

File: requirements/extra.txt
Comment (tchaton) [inline requirements/extra.txt:11]: I think we will have to make a fork and merge sendrecv with master, so both our PR works.

File: pytorch_lightning/trainer/connectors/precision_connector.py
Comment (Borda) [reply pytorch_lightning/trainer/connectors/precision_connector.py:72]: at this point, it is just string...

File: pytorch_lightning/plugins/sharded_plugin.py
Comment (rohitgr7) [reply pytorch_lightning/plugins/sharded_plugin.py:48]: it's not merged yet. Waiting for review :] https://github.com/PyTorchLightning/pytorch-lightning/pull/3671

File: pytorch_lightning/plugins/sharded_plugin.py
Comment (SeanNaren) [reply pytorch_lightning/plugins/sharded_plugin.py:56]: Rank zero is the only rank that contains the optim state

File: pytorch_lightning/trainer/connectors/precision_connector.py
Comment (SeanNaren) [reply pytorch_lightning/trainer/connectors/precision_connector.py:72]: It doesn't, and it isn't supported by fairscale yet since apex has the same issue with scaling gradients. Might be worth adding a check to ensure the user doesn't use both at the same time!

File: pytorch_lightning/plugins/sharded_plugin.py
Comment (SeanNaren) [reply pytorch_lightning/plugins/sharded_plugin.py:43]: I can align the name with the refactor however this hook is separate to the model (exists within the plugin)

File: requirements/extra.txt
Comment (SeanNaren) [reply requirements/extra.txt]: I think you should handle that within the pipe PR, just override this requirement once merged

File: pytorch_lightning/plugins/sharded_plugin.py
Comment (SeanNaren) [reply pytorch_lightning/plugins/sharded_plugin.py:43]: This is out of scope, since I'll need to modify across accelerators. We can do this after refactors have been merged (trying to keep this to the core feature only)

File: pytorch_lightning/plugins/sharded_plugin.py
Comment (SeanNaren) [reply pytorch_lightning/plugins/sharded_plugin.py:26]: Let's figure this out in a future PR (out of scope)

File: pytorch_lightning/trainer/connectors/precision_connector.py
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/precision_connector.py:99]: Is it the right place for `MisconfigurationException` ? I would have excepted this to be raised while creating `DDPShardedPlugin` already.

File: benchmarks/test_sharded_parity.py
Comment (Borda) [inline benchmarks/test_sharded_parity.py:91]: how this `os.getenv("PL_RUNNING_SPECIAL_TESTS", '0')` works if you patch it above with `@mock.patch.dict(os.environ, {"PL_DEV_DEBUG": "1"})`?

File: benchmarks/test_sharded_parity.py
Comment (Borda) [inline benchmarks/test_sharded_parity.py:245]:
```suggestion
        model_cls: LightningModule,
```

File: benchmarks/test_sharded_parity.py
Comment (Borda) [inline benchmarks/test_sharded_parity.py:251]: 25% is quite a lot...

File: pytorch_lightning/overrides/fairscale.py
Comment (Borda) [inline pytorch_lightning/overrides/fairscale.py:32]: shall here be also else and set `LightningShardedDataParallel`
```suggestion
            return outputs
else:
    LightningShardedDataParallel = ...
```

File: pytorch_lightning/plugins/sharded_native_amp_plugin.py
Comment (Borda) [inline pytorch_lightning/plugins/sharded_native_amp_plugin.py:22]:
```suggestion
from pytorch_lightning.plugins.native_amp import NativeAMPPlugin

if NATIVE_AMP_AVAILABLE and FAIRSCALE_AVAILABLE:
    from fairscale.optim import OSS
    from fairscale.optim.grad_scaler import ShardedGradScaler
```
just formating

File: pytorch_lightning/trainer/connectors/precision_connector.py
Comment (Borda) [inline pytorch_lightning/trainer/connectors/precision_connector.py:20]:
```suggestion
from pytorch_lightning.utilities import (
    APEX_AVAILABLE, NATIVE_AMP_AVAILABLE, AMPType, rank_zero_warn, FAIRSCALE_AVAILABLE)
```

File: pytorch_lightning/trainer/connectors/precision_connector.py
Comment (Borda) [inline pytorch_lightning/trainer/connectors/precision_connector.py:23]:
```suggestion
```

File: pytorch_lightning/trainer/connectors/precision_connector.py
Comment (Borda) [inline pytorch_lightning/trainer/connectors/precision_connector.py:78]:
```suggestion
                raise MisconfigurationException(
                    'Sharded Plugin is not supported with Apex AMP, please using native AMP for 16-bit precision.'
                )
```

File: tests/plugins/test_sharded_plugin.py
Comment (Borda) [inline tests/plugins/test_sharded_plugin.py:301]: why all these returns?

**Review by Borda — APPROVED** [review]
lgtm

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
the addition of the plugins LGTM, nice work @SeanNaren ! (I only looked at the code, don't have much time to test it on my machine sorry)
The only concern I have is about the testing, which adds 3mins (11 mins on master vs. 14 mins here) and I am not sure if the addition of the sharded feature justifies this overhead. Are my observations correct? 
Not sure how feasible this is here, but I would suggest to target more granular unit testing that does not involve a full expensive training loop.

**Pull request merged**
